### PR TITLE
Feature/SOF-7476 update: rewrite Standata to new specs

### PR DESCRIPTION
**Pull request by VsevolodX**

### CODE REVIEW

File: src/py/mat3ra/standata/base.py [hunk 1]
```diff
@@ -1,88 +1,40 @@
...
 
-EntityConfig = TypedDict("EntityConfig", {"categories": Dict[str, List[str]], "entities": List[EntityItem]})
+class StandataEntity(BaseModel):
+    filename: str
+    categories: List[str]
 
 
-class Standata:
-    """The Standata class associates the entity data files with categories and allows for tag-based queries.
+CATEGORY_SEPARATOR = "/"
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/base.py:13]: Place above the first class definition

File: src/py/mat3ra/standata/base.py [hunk 2]
```diff
@@ -92,68 +44,186 @@ def convert_tag_to_category(self, *tags: str):
             'electrical_conductivity' and 'type'. This function returns all occurrences of a tag as
             '<category_type>/<tag>'.
         """
-        return [cf for cf in self.categories if any([cf.split("/")[1] == t for t in tags])]
+        return [
+            cf for cf in self.get_categories_as_list() if any((cf.split(CATEGORY_SEPARATOR)[-1] == t) for t in tags)
+        ]
+
+    def get(self, key: str, default=None):
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/base.py:51]: We shouldn't need it or it should be called `get_from_categories`

File: src/py/mat3ra/standata/base.py [hunk 3]
```diff
@@ -92,68 +44,186 @@ def convert_tag_to_category(self, *tags: str):
...
+        filenames = []
+        for entity in self.entities:
+            if re.search(regex, entity.filename):
+                filenames.append(entity.filename)
+        return filenames
 
-    def __create_table(self) -> pd.DataFrame:
-        """Creates lookup table for filenames and associated categories.
+    @property
+    def __lookup_table(self) -> pd.DataFrame:
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/base.py:113]: Put a TODO note explaining that this is not used, but left in preparation for the future when increasing the number of materials

File: src/py/mat3ra/standata/base.py [hunk 4]
```diff
@@ -92,68 +44,186 @@ def convert_tag_to_category(self, *tags: str):
...
+        super().__init__(filesMapByName=self._initialize_files_map(data), standataConfig=self._initialize_config(data))
 
     @staticmethod
-    def load_entity(filepath: Path) -> Optional[dict]:
-        """Loads entity config from file (JSON).
+    def _initialize_files_map(data: Dict) -> StandataFilesMapByName:
+        """
+        Initialize the StandataFilesMapByName from the input data.
+        """
+        return StandataFilesMapByName(dictionary=data.get("filesMapByName", {}))
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/base.py:172]: I'd split this into 2 lines for clarity like you have it on line 179

File: src/py/mat3ra/standata/base.py [hunk 5]
```diff
@@ -92,68 +44,186 @@ def convert_tag_to_category(self, *tags: str):
...
+
+    @staticmethod
+    def _initialize_config(data: Dict) -> StandataConfig:
+        """
+        Initialize StandataConfig from the input data.
+        """
+        config_data = data.get("standataConfig", {})
+        return StandataConfig(
+            categories=config_data.get("categories", {}),
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/base.py:181]: Same as above - let's define `categories = config_data.get("categories", {})` and entities above on separate lines

File: src/py/mat3ra/standata/build/builder.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+from ..base import StandataConfig
+
+EntityItem = TypedDict("EntityItem", {"filename": str, "categories": List[str]})
+
+EntityConfig = TypedDict("EntityConfig", {"categories": Dict[str, List[str]], "entities": List[EntityItem]})
+
+
+class StandataBuilder:
+    """The Standata class associates the entity data files with categories and allows for tag-based queries.
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/build/builder.py:15]: This line needs to be replaced

File: src/py/mat3ra/standata/build/builder.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import yaml
+from pathlib import Path
+from typing import Dict, List, Optional, TypedDict, Union
+
+
+from ..base import StandataConfig
+
+EntityItem = TypedDict("EntityItem", {"filename": str, "categories": List[str]})
+
+EntityConfig = TypedDict("EntityConfig", {"categories": Dict[str, List[str]], "entities": List[EntityItem]})
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/build/builder.py:11]: I don't think these should be used

File: src/py/mat3ra/standata/build/cli.py [hunk 8]
```diff
@@ -2,29 +2,31 @@
...
-        entity_path = std.entity_dir / entity["filename"]
+    for entity in standata_config.entities:
+        categories = standata_config.convert_tags_to_categories_list(*entity.categories)
+        entity_path = entity_path_parent / entity.filename
 
         for category in categories:
             category_dir = categories_root / category
             category_dir.mkdir(parents=True, exist_ok=True)
-            linked_entity = category_dir / entity["filename"]
+            linked_entity = category_dir / entity.filename
```
Comment (timurbazhirov) [inline src/py/mat3ra/standata/build/cli.py:29]: Does this work?

File: tests/py/unit/materials/test_standata_init.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+from mat3ra.standata.materials import Materials
```
Comment (timurbazhirov) [inline tests/py/unit/materials/test_standata_init.py:1]: We should combine this file and test_materials_data into test_materials_data

Comment (timurbazhirov) [reply tests/py/unit/materials/test_standata_init.py:1]: We can make another file for test_standata_build

Comment (VsevolodX) [reply tests/py/unit/materials/test_standata_init.py]: added Claude generated tests for builder and CLI

File: tests/py/unit/materials/test_standata_build.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    for category in categories:
+        category_dir = categories_root / category
+        category_dir.mkdir(parents=True, exist_ok=True)
+        linked_entity = category_dir / entity.filename
+        if not linked_entity.exists():
+            try:
+                linked_entity.symlink_to(entity_path)
+            except PermissionError:
+                pytest.skip("No permission to create symlinks")
```
Comment (timurbazhirov) [inline tests/py/unit/materials/test_standata_build.py:149]: Lines 101-120 and 132-149 need to be reused from above cli.py

**Pull request merged**